runtime/src/snapshot_utils.rs

```diff
@@ -478,37 +479,47 @@ pub fn move_and_async_delete_path_contents(path: impl AsRef<Path>) {
 }
 
 /// Delete directories/files asynchronously to avoid blocking on it.
-/// First, in sync context, rename the original path to *_deleted,
-/// then spawn a thread to delete the renamed path.
-/// If the process is killed and the deleting process is not done,
-/// the leftover path will be deleted in the next process life, so
-/// there is no file space leaking.
+/// First, in sync context, check if the original path exists, if it
+/// does, rename the original path to *_to_be_deleted.
+/// If there's an in-progress deleting thread for this path, return.
+/// Then spawn a thread to delete the renamed path.
 pub fn move_and_async_delete_path(path: impl AsRef<Path>) {
-    let mut path_delete = PathBuf::new();
-    path_delete.push(&path);
-    path_delete.set_file_name(format!(
-        "{}{}",
-        path_delete.file_name().unwrap().to_str().unwrap(),
-        "_to_be_deleted"
-    ));
+    lazy_static! {
+        static ref IN_PROGRESS_DELETES: Mutex<HashSet<PathBuf>> = Mutex::new(HashSet::new());
+    };
 
-    if path_delete.exists() {
-        std::fs::remove_dir_all(&path_delete).unwrap();
-    }
+    // Grab the mutex so no new async delete threads can be spawned for this path.
+    let mut lock = IN_PROGRESS_DELETES.lock().unwrap();
 
+    // If the path does not exist, there's nothing to delete.
     if !path.as_ref().exists() {
         return;
     }
 
+    // If the original path (`path_delete` here) is already being deleted,
+    // then the path should not be moved and deleted again.
+    let mut path_delete = path.as_ref().to_path_buf();
+    if lock.contains(&path_delete) {
+        return;
+    }
+
+    path_delete.push("_to_be_deleted");
     if let Err(err) = std::fs::rename(&path, &path_delete) {
```

Comment:
if we can't rename it - there's a few reasons:
1. permission issues (can't do anything about that 😦 )
2. `path_delete` exists
3. `path` does not exist (already confirmed it does, while holding the lock)

if `path_delete` exists, then we have both `path` and `path_exists`, potentially because of a crash mid-delete on the last validator run.

Reply:
another case is that we have `move_and_async_delete_path_contents` which re-creates the original directory!